scipy/linalg/_generate_pyx.py

```diff
@@ -59,13 +68,46 @@ def arg_names_and_types(args):
              'zselect1': '_zselect1', 'zselect2': '_zselect2'}
 
 
-def arg_casts(arg):
-    if arg in ['npy_complex64', 'npy_complex128', '_cselect1', '_cselect2',
+def arg_casts(argname, argtype, index):
+    if argtype in complex_types:
+        # If the argument type is a complex number, we do the following:
+        # 1. Create a union with two members: one ptr to native complex number and a ptr to npy_complex64/128
+        # 2. We assign the native argument to the corresponding union field
+        # 3. Pass the npy_complex to the wrapped Fortran subroutine
```

Comment:
Can we mention something here that this works because in all cases the union members have the same memory layout, and we want to avoid direct casts for $REASON and just reinterpret the bits directly?

Bonus: Linus on using [unions](https://lore.kernel.org/lkml/CAHk-=wj9guryjifHyr26w73ta+kNeoHtGde682Z5N6OSjKu4UQ@mail.gmail.com/)

Reply:
It doesn't apply here. We are changing the containing type name not interpreting bits. Both are ordered floats for real and complex parts.

Reply:
If it doesn't apply, why do we need a union in the first place, rather than casting? I haven't looked in detail, but I assumed the types are not inter-convertible.

> Also these are not pointers so not sure about Torvalds quote.

AFAIU, "proper" use of unions is fundamentally about overcoming/circumventing type incompatibility, and has nothing to do with pointers per se.

Reply:
PS. Citing Linus was not meant to be lecturing anyone about anything. I just remembered having seen that comment recently and this PR reminded me of it.

Reply:
> We are changing the containing type name not interpreting bits.

Correct. We're doing that to remove compiler warnings. Compilers warn about incompatible types even though the memory layout is exactly the same.

Reply:
I can always add a comment explaining that, if it makes things clearer.

Reply:
> Compilers warn about incompatible types even though the memory layout is exactly the same.

Which is still broadly what I was aiming at. In this case we know better than the compiler, and we use the union to silence that warning intentionally (this much is parallel with what I had quoted).

> Because otherwise we have to cast back and forth in every usage of NumPy functions

That's fair, but should ideally be mentioned in a comment as well, IMO.

Reply:
> Because otherwise we have to cast back and forth in every usage of NumPy functions

Not sure, but I think that Cython won't even let us cast back and forth.

Reply:
The reason I brought it up is because it's (IMO) not obvious at a glance what's happening, unless the reader is already familiar with that (rather less common) technique. I don't doubt that you both know about it, but wanted to preempt the "wait, what's going on here" of a less experienced reader, and make it explicit that "this is what's happening, it's intentional".

Reply:
The clearer the better so no objections to more text

Reply:
Done in https://github.com/scipy/scipy/pull/18403/commits/ce3009cb5dd689f190270f2149955252c5c2657c. Let me know if I should include something else in it. Thanks for the help!